File: src/org/cappuccino_project/ide/intellij/plugin/psi/impl/ObjJResolveUtil.java
package org.cappuccino_project.ide.intellij.plugin.psi.impl;

import com.intellij.psi.PsiElement;
import com.intellij.psi.ResolveState;
import com.intellij.psi.scope.PsiScopeProcessor;
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJCompositeElement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;


public class ObjJResolveUtil {
    private ObjJResolveUtil() {}


    public static boolean treeWalkUp(@Nullable PsiElement place, @NotNull PsiScopeProcessor processor) {
        PsiElement lastParent = null;
        PsiElement run = place;
        while (run != null) {
            if (place != run && !run.processDeclarations(processor, ResolveState.initial(), lastParent, place)) return false;
            lastParent = run;
            run = run.getParent();
        }
        return true;
    }

    public static boolean processChildren(@NotNull PsiElement element,
                                          @NotNull PsiScopeProcessor processor,
                                          @NotNull ResolveState substitutor,
                                          @Nullable PsiElement lastParent,
                                          @NotNull PsiElement place) {
        PsiElement run = lastParent == null ? element.getLastChild() : lastParent.getPrevSibling();
        while (run != null) {
            if (run instanceof ObjJCompositeElement && !run.processDeclarations(processor, substitutor, null, place)) return false;
            run = run.getPrevSibling();
        }
        return true;
    }
}

File: src/cappuccino/ide/intellij/plugin/psi/impl/ObjJResolveUtil.kt
package cappuccino.ide.intellij.plugin.psi.impl

import cappuccino.ide.intellij.plugin.psi.interfaces.ObjJCompositeElement
import com.intellij.psi.PsiElement
import com.intellij.psi.ResolveState
import com.intellij.psi.scope.PsiScopeProcessor


object ObjJResolveUtil {
    fun processChildren(element: PsiElement,
                        processor: PsiScopeProcessor,
                        substitutor: ResolveState,
                        lastParent: PsiElement?,
                        place: PsiElement): Boolean {
        var run: PsiElement? = if (lastParent == null) element.lastChild else lastParent.prevSibling
        while (run != null) {
            if (run is ObjJCompositeElement && !run.processDeclarations(processor, substitutor, null, place)) return false
            run = run.prevSibling
        }
        return true
    }
}